gui/journal.lua

```diff
@@ -0,0 +1,797 @@
+                self.render_start_line = self.editor.cursor.y - self.editor.frame_body.height + 1
+            end
+
+            self:updateLayout()
+            if self.on_change then
+                self.on_change(val)
+            end
+        end
+    }
+
+    self:addviews{
+        self.scrollbar,
+        self.editor
+    }
+    self:setFocus(true)
+end
+
+function TextEditor:getPreferredFocusState()
+    return true
+end
+
+function TextEditor:postUpdateLayout()
+    self:updateScrollbar()
+end
+
+function TextEditor:onScrollbar(scroll_spec)
+    local height = self.editor.frame_body.height
+    if scroll_spec == 'down_large' then
+        self.render_start_line = self.render_start_line + math.ceil(height / 2)
+    elseif scroll_spec == 'up_large' then
+        self.render_start_line = self.render_start_line - math.ceil(height / 2)
+    elseif scroll_spec == 'down_small' then
+        self.render_start_line = self.render_start_line + 1
+    elseif scroll_spec == 'up_small' then
+        self.render_start_line = self.render_start_line - 1
+    else
+        self.render_start_line = tonumber(scroll_spec)
+    end
+
+    self.render_start_line = math.min(
+        #self.editor.lines - height + 1,
+        math.max(1, self.render_start_line)
+    )
+
+    self:updateScrollbar()
+    -- local max_page_top = math.max(1, #self.choices - self.page_size + 1)
+    -- self.page_top = math.max(1, math.min(max_page_top, self.page_top + v))
+    -- update_list_scrollbar(self)
+end
+
+function TextEditor:updateScrollbar()
+    local lines_count = #self.editor.lines
+
+    self.scrollbar:update(
+        self.render_start_line,
+        self.frame_body.height,
+        lines_count
+    )
+    if (self.frame_body.height >= lines_count) then
+        self.render_start_line = 1
+    end
+end
+
+function TextEditor:onInput(keys)
+    if self.scrollbar:onInput(keys) then
+        return true
+    end
+
+    if not self.scrollbar:getMousePos() and not self.scrollbar.is_dragging then
+        return self.editor:onInput(keys)
+    end
+end
+
+function TextEditor:renderSubviews(dc)
+    dc:clear()
+    self.editor.frame_body.y1 = self.frame_body.y1-(self.render_start_line - 1)
+    self.editor:render(dc)
+    self.scrollbar:render(dc)
+end
+
+-- multiline text editor, features
+--[[
+Supported features:
+ - cursor controlled by arrow keys (left, right, top, bottom)
+ - fast rewind by shift+left/ctrl+b and shift+right/ctrl+f
+ - remember longest x for up/bottom cursor movement
+ - mouse control for cursor
+ - support few new lines (submit key)
+ - wrapable text
+ - backspace
+ - ctrl+d as delete
+ - ctrl+a / ctrl+e go to beginning/end of line
+ - ctrl+u delete current line
+ - ctrl+w delete last word
+ - mouse text selection and replace/remove features for it
+ - local copy/paste selection text or current line (ctrl+x/ctrl+c/ctrl+v)
+ - go to text begin/end by shift+up/shift+down
+--]]
```

Comment:
this is already documented in the .rst file -- this copy will likely just get out of date and should be removed